xlators/mgmt/glusterd/src/glusterd-volgen.c

```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
     return ret;
 }
 
+gf_boolean_t
+check_user_xlator_position(dict_t *dict, char *key, data_t *value,
+                           void *prev_xlname)
+{
+    if (strncmp(key, "user.xlator.", strlen("user.xlator.")) != 0) {
+        return false;
+    }
+
+    if (fnmatch("user.xlator.*.*", key, 0) == 0) {
+        return false;
+    }
+
+    char *value_str = data_to_str(value);
+    if (!value_str) {
+        return false;
+    }
+
+    if (strcmp(data_to_str(value), prev_xlname) == 0) {
+        gf_log("glusterd", GF_LOG_INFO,
+               "found insert position of user-xlator(%s)", key);
+        return true;
+    }
+
+    return false;
+}
+
+int
+set_user_xlator_option(dict_t *set_dict, char *key, data_t *value, void *data)
+{
+    xlator_t *xl = data;
+    char *optname = strrchr(key, '.') + 1;
+
+    gf_log("glusterd", GF_LOG_DEBUG, "set user xlator option %s = %s", key,
+           value->data);
+
+    return xlator_set_option(xl, optname, strlen(optname), data_to_str(value));
+}
+
+int
+insert_user_xlator_to_graph(dict_t *set_dict, char *key, data_t *value,
+                            void *action_data)
+{
+    int ret = -1;
+
+    struct {
+        volgen_graph_t *graph;
+        char *volname;
+    } *data = action_data;
+
+    char *xlator_name = strrchr(key, '.') + 1;  // user.xlator.<xlator_name>
+    char *xlator_option_matcher = NULL;
+    char *type = NULL;
+    xlator_t *xl = NULL;
+
+    // convert optkey to xlator type
+    if (gf_asprintf(&type, "user/%s", xlator_name) == -1) {
+        gf_log("glusterd", GF_LOG_ERROR, "failed to generate user-xlator type");
+        goto out;
+    }
+
+    gf_log("glusterd", GF_LOG_INFO, "add user xlator=%s to graph", type);
+
+    xl = volgen_graph_add(data->graph, type, data->volname);
+    if (!xl) {
+        goto out;
+    }
+
+    ret = gf_asprintf(&xlator_option_matcher, "user.xlator.%s.*", xlator_name);
+    if (ret == -1) {
+        gf_log("glusterd", GF_LOG_ERROR,
+               "failed to generate user-xlator option matcher");
+        goto out;
+    }
+
+    dict_foreach_fnmatch(set_dict, xlator_option_matcher,
+                         set_user_xlator_option, xl);
+
+out:
+    if (type)
+        GF_FREE(type);
+
+    return ret;
+}
+
+int
+check_and_add_user_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+                      char *prev_xlname)
+{
+    if (!prev_xlname)
+        goto out;
+
+    struct {
+        volgen_graph_t *graph;
+        char *volname;
+    } data;
+
+    data.graph = graph;
+    data.volname = volname;
+
+    if (dict_foreach_match(set_dict, check_user_xlator_position, prev_xlname,
+                           insert_user_xlator_to_graph, &data) < 0) {
+        return 1;
+    }
+
+out:
+    return 0;
```

Comment:
> Is it possible to add some validation if the xlator insertion was unsuccessful? We currently seem to be returning success even if an invalid prev_xlname was given at the gluster volume set CLI.

Currently, it looks like inserting a debug xlator is not validating.
If set an invalid xlator name to debug.trace, the CLI succeeds and it is recorded in volinfo, but the debug.trace xlator is not inserted in the graph.

```
sh-4.4$ sudo gluster vol set gvol debug.trace unknown-xlator
volume set: success
sh-4.4$ sudo grep 'debug.trace' /var/lib/glusterd/vols/gvol/info
debug.trace=unknown-xlator
```

Should I do a strict validation of input xlname?

If you want to perform validation, we need to have a server xlator list in both CLI and glusterd.
For a better implementation, I think we need to define it in a .h file and share it between cli and glusterd.
And that would be a big fix.

Reply:
That sounds complex. Would something like this work? (I haven't tested it)
```
check_user_xlator_position()
{
    if (set_dict does not contain the key "user.xlator.*")
        return 0;

     if (value == prev_xl_name) {
        insert_user_xlator_to_graph();
        return 0 if successful;
    }
        
     return -EINVAL //value did not match a valid xlator name in the graph.   
        
}
```

Reply:
I afraid such a code will not work.
set_dict has key-value pairs for all options in volinfo, and server_graph_builder() loops through the hardcoded server xlator list and checks if opt-value matches the xlator name in it.
So if returns an error when opt value does not match, it will not work.

Reply:
Okay, will go through the current patch set soon.

Reply:
I pushed a new commit that changes the volgen process to check if opt-value is invalid, and fail to generate volfile if it is invalid.